types/wlr_xdg_foreign_v2.c

```diff
@@ -0,0 +1,411 @@
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static bool verify_is_toplevel(struct wl_resource *client_resource,
+		struct wlr_surface *surface) {
+	if (wlr_surface_is_xdg_surface(surface)) {
+		struct wlr_xdg_surface *xdg_surface =
+			wlr_xdg_surface_from_wlr_surface(surface);
+		if (xdg_surface->role != WLR_XDG_SURFACE_ROLE_TOPLEVEL) {
+			wl_resource_post_error(client_resource, -1,
+					"surface must be an xdg_toplevel");
+			return false;
+		}
+	} else {
+		wl_resource_post_error(client_resource, -1,
+				"surface must be an xdg_surface");
+		return false;
+	}
+	return true;
+}
+
+static void destroy_imported_child(struct wlr_xdg_imported_child_v2 *child) {
+	wl_list_remove(&child->xdg_toplevel_set_parent.link);
+	wl_list_remove(&child->xdg_surface_unmap.link);
+	wl_list_remove(&child->link);
+	free(child);
+}
+
+static void handle_child_xdg_surface_unmap(
+		struct wl_listener *listener, void *data) {
+	struct wlr_xdg_imported_child_v2 *child =
+		wl_container_of(listener, child, xdg_surface_unmap);
+	destroy_imported_child(child);
+}
+
+static void handle_xdg_toplevel_set_parent(
+		struct wl_listener *listener, void *data) {
+	struct wlr_xdg_imported_child_v2 *child =
+		wl_container_of(listener, child, xdg_toplevel_set_parent);
+	destroy_imported_child(child);
+}
+
+static void xdg_imported_handle_set_parent_of(struct wl_client *client,
+		struct wl_resource *resource,
+		struct wl_resource *child_resource) {
+	struct wlr_xdg_imported_v2 *imported =
+		xdg_imported_from_resource(resource);
+	if (imported->exported == NULL) {
+		return;
+	}
+	struct wlr_surface *wlr_surface = imported->exported->surface;
+	struct wlr_surface *wlr_surface_child =
+		wlr_surface_from_resource(child_resource);
+
+	if (!verify_is_toplevel(resource, wlr_surface_child)) {
+		return;
+	}
+	if (wlr_surface_is_xdg_surface(wlr_surface)
+			!= wlr_surface_is_xdg_surface(wlr_surface_child)) {
+		wl_resource_post_error(resource, -1,
+				"surfaces must have the same role");
+		return;
+	}
+	struct wlr_xdg_imported_child_v2 *child;
+	wl_list_for_each(child, &imported->children, link) {
+		if (child->surface == wlr_surface_child) {
+			return;
+		}
+	}
+
+	child = calloc(1, sizeof(struct wlr_xdg_imported_child_v2));
+	if (child == NULL) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+	child->surface = wlr_surface_child;
+	child->xdg_surface_unmap.notify = handle_child_xdg_surface_unmap;
+	child->xdg_toplevel_set_parent.notify = handle_xdg_toplevel_set_parent;
+
+	if (wlr_surface_is_xdg_surface(wlr_surface)) {
+		struct wlr_xdg_surface *surface =
+			wlr_xdg_surface_from_wlr_surface(wlr_surface);
+		struct wlr_xdg_surface *surface_child =
+			wlr_xdg_surface_from_wlr_surface(wlr_surface_child);
+
+		wlr_xdg_toplevel_set_parent(surface_child, surface);
+		wl_signal_add(&surface_child->events.unmap,
+				&child->xdg_surface_unmap);
+		wl_signal_add(&surface_child->toplevel->events.set_parent,
+				&child->xdg_toplevel_set_parent);
+	}
```

Comment:
`child` is leaked in the else branch, because `wlr_xdg_toplevel_set_parent` won't free it in `xdg_imported_handle_resource_destroy`.

Reply:
if we remove the if() there is no else, right?